File: Week4 (Sort, Search, Tree-1)/Search/khj_BST.py
class TreeNode():
    def __init__(self, x:int):
        self.val = x
        self.left = None
        self.right = None
        self.parent = None

class BST():
    def __init__(self):
        self.root = None
    
    
    #search
    def __searchHelp(self, curNode: TreeNode, x: int) -> TreeNode:
        if not curNode:
            return None
        if x == curNode.val:
            return curNode
        elif x < curNode.val:
            return self.__searchHelp(curNode.left, x)
        else:
            return self.__searchHelp(curNode.right, x)
    
    def search(self, x: int):
        return self.__searchHelp(self.root, x)
    

    #insert
    def __insertHelp(self, curNode: TreeNode, x: int) -> TreeNode:
        if not curNode:
            return TreeNode(x)
        if x < curNode.val:
            if curNode.left is None:
                curNode.left = TreeNode(x)
                curNode.left.parent = curNode
            else:
                curNode.left = self.__insertHelp(curNode.left, x)
        elif x > curNode.val:
            if curNode.right is None:
                curNode.right = TreeNode(x)
                curNode.right.parent = curNode
            else:
                curNode.right = self.__insertHelp(curNode.right, x)
        return curNode

    def insert(self, x: int):
        if self.root is None:
            self.root = TreeNode(x)
        else:
            self.root = self.__insertHelp(self.root, x)


    #delete
    def __findMin(self, node: TreeNode) -> TreeNode:
        while node.left:
            node = node.left
        return node

    def __deleteHelp(self, node: TreeNode):
        if not node:
            return None
        
        # Node has no children
        if node.left is None and node.right is None:
            if node == self.root:
                self.root = None
            elif node == node.parent.left:
                node.parent.left = None
            else:
                node.parent.right = None
        
        # Node has one child
        elif not node.left or not node.right:
            child = node.left if node.left else node.right
            if node == self.root:
                self.root = child
            elif node == node.parent.left:
                node.parent.left = child
            else:
                node.parent.right = child
            if child:
                child.parent = node.parent

        # Node has two children    
        else:
            replace = self.__findMin(node.right)
            node.val = replace.val
            self.__deleteHelp(replace)

    def delete(self, x: int):
        node = self.search(x)
        if node:
            self.__deleteHelp(node)

# Main function to demonstrate usage
def main():
    bst = BST()
    
    # Insert nodes into the BST
    nodes_to_insert = [20, 10, 30, 5, 15, 25, 35]
    for value in nodes_to_insert:
        bst.insert(value)
        print(f"Inserted {value} into the BST.")
    
    # Search for a node in the BST
    search_value = 15
    search_result = bst.search(search_value)
    if search_result:
        print(f"Found node with value {search_value}.")
    else:
        print(f"Node with value {search_value} not found.")
    
    # Delete nodes from the BST
    nodes_to_delete = [10, 30, 20]  # Delete some nodes to demonstrate different cases
    for value in nodes_to_delete:
        bst.delete(value)
        print(f"Deleted node with value {value} from the BST.")
    
    # Final tree status (in-order traversal for visualization)
    def inorder_traversal(node):
        if node:
            inorder_traversal(node.left)
            print(node.val, end=' ')
            inorder_traversal(node.right)
    
    print("Final BST (in-order traversal):")
    inorder_traversal(bst.root)
    print()

if __name__ == "__main__":
    main()

File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/BST/khj_BST.c
#include <stdio.h>
#include <stdlib.h>

typedef struct treeNode{
    int val;
    struct treeNode* left;
    struct treeNode* right;
} TreeNode;

TreeNode *create_node(int x) {
    TreeNode * new_node = (TreeNode*)malloc(sizeof(TreeNode));
    new_node->val = x;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}


//search
TreeNode *search(TreeNode *node, int key) {
    if (node == NULL) return NULL;
    // if (key == node->val) return node;
    if (key  < node->val) {
        return search(node->left, key);
    }
    else if (key > node->val) {
        return search(node->right, key);
    }
    return node;
}

TreeNode *insert(TreeNode *node, int key) {
    if (node == NULL) {
        return create_node(key);
    }

    if (key < node-> val) {
        node->left = insert(node->left, key);
    }
    else {
        node->right = insert(node->right ,key);
    }
    return node;
}

// //insert
// TreeNode *insert(TreeNode *node, int key) {
//     if (node == NULL) return NULL;
//     if (key < node-> val) {
//         if (node->left == NULL) {
//             node->left->val = key;
//             node->left->parent = node;
//         }
//         else {
//             node->left = insert(node->left, key);
//         }
//     }

//     else if (key > node->val) {
//         if (node->right == NULL) {
//             node->right->val = key;
//             node->right->parent = node;
//         }
//         else {
//             node->right = insert(node->right, key);
//         }
//     }
//     return node;
// }

//delete
TreeNode* findMin(TreeNode* node) {
    while (node->left) {
        node = node->left;
    }
    return node;
} 

TreeNode* delete(TreeNode* node, int key) {
    if (node == NULL) return NULL;
    
    if (node->val == key) {

        //Node has no children
        if (node->left == NULL && node->right == NULL) {
            node = NULL;
        }

        //Node has 1 child
        else if (node->left != NULL && node->right == NULL) {
            node = node->left;
        }

        else if (node->left == NULL && node->right != NULL) {
            node = node->right;
        }

        //Node has 2 children
        else {
            TreeNode* temp = findMin(node->right);
            node->val = temp->val;
            node->right = delete(node->right, temp->val);
            
        }
    }
    else if (node->val < key) {
        node->right = delete(node->right, key);
    }
    else if (node->val > key) {
        node->left =  delete(node->left, key);
    }
    // return node;

}

//Modify
TreeNode* modify(TreeNode *node, int data, int newdata) {
    if (node == NULL) return;
    if (node->val == data) {
        node->val = newdata;
    }
    else if (data < node->val) {
        modify(node->left, data, newdata);
    }
    else {
        modify(node->right, data, newdata);
    }
    return node;
}


void inorder(TreeNode* node) {
        if (node != NULL) {
            inorder(node->left);
            printf("%d ", node->val);
            inorder(node->right);
        }
    }


int main() {
    TreeNode* root = NULL;

    // Insert nodes
    root = insert(root, 50);
    insert(root, 30);
    insert(root, 20);
    insert(root, 40);
    insert(root, 70);
    insert(root, 60);
    insert(root, 80);
    

    
    printf("Inorder traversal of the tree: ");
    inorder(root);
    printf("\n");

    
    // Search for a value
    TreeNode* result = search(root, 40);
    if (result != NULL) {
        printf("Found value 40 in the tree.\n");
    } else {
        printf("Value 40 not found in the tree.\n");
    }

    // Modify a value
    printf("Modifying value 40 to 45.\n");
    modify(root, 40, 45);
    printf("Inorder traversal after modification: ");
    inorder(root);
    printf("\n");


    // Delete a node
    printf("Deleting node with value 30 and 50.\n");
    delete(root, 30);
    delete(root, 50);
    // printf("Deleting node with value 20: \n");
    // delete(root, 20);
    printf("Inorder traversal after deletion: ");
    
    inorder(root);
    printf("\n");

    return 0;
}